api/args/reddit/search/search.go

```diff
@@ -0,0 +1,192 @@
+package search
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/masa-finance/tee-worker/api/args/base"
+	"github.com/masa-finance/tee-worker/api/types"
+)
+
+var (
+	ErrInvalidType       = errors.New("invalid type")
+	ErrInvalidSort       = errors.New("invalid sort")
+	ErrTimeInTheFuture   = errors.New("after field is in the future")
+	ErrNoQueries         = errors.New("queries must be provided for all query types except scrapeurls")
+	ErrNoUrls            = errors.New("urls must be provided for scrapeurls query type")
+	ErrQueriesNotAllowed = errors.New("the scrapeurls query type does not admit queries")
+	ErrUrlsNotAllowed    = errors.New("urls can only be provided for the scrapeurls query type")
+	ErrUnmarshalling     = errors.New("failed to unmarshal reddit search arguments")
+)
+
+const (
+	// These reflect the default values in https://apify.com/trudax/reddit-scraper/input-schema
+	DefaultMaxItems       = 10
+	DefaultMaxPosts       = 10
+	DefaultMaxComments    = 10
+	DefaultMaxCommunities = 2
+	DefaultMaxUsers       = 2
+	DefaultSort           = types.RedditSortNew
+)
+
+const DomainSuffix = "reddit.com"
+
+// Verify interface implementation
+var _ base.JobArgument = (*Arguments)(nil)
+
+// Arguments defines args for Reddit scrapes
+// see https://apify.com/trudax/reddit-scraper
+type Arguments struct {
+	Type           types.Capability     `json:"type"`
+	Queries        []string             `json:"queries"`
+	URLs           []string             `json:"urls"`
+	Sort           types.RedditSortType `json:"sort"`
+	IncludeNSFW    bool                 `json:"include_nsfw"`
+	SkipPosts      bool                 `json:"skip_posts"`      // Valid only for searchusers
+	After          time.Time            `json:"after"`           // valid only for scrapeurls and searchposts
+	MaxItems       uint                 `json:"max_items"`       // Max number of items to scrape (total), default 10
+	MaxResults     uint                 `json:"max_results"`     // Max number of results per page, default MaxItems
+	MaxPosts       uint                 `json:"max_posts"`       // Max number of posts per page, default 10
+	MaxComments    uint                 `json:"max_comments"`    // Max number of comments per page, default 10
+	MaxCommunities uint                 `json:"max_communities"` // Max number of communities per page, default 2
+	MaxUsers       uint                 `json:"max_users"`       // Max number of users per page, default 2
+	NextCursor     string               `json:"next_cursor"`
+}
+
+func (t *Arguments) UnmarshalJSON(data []byte) error {
```

Comment:
...and this sort of thing is why we ask "Why doesn't Go have metaprogramming"? Good that we're using the same pattern everywhere!

I wonder... should we have a document (perhaps in the README or somewhere else) where we document the patterns that we use and, in cases like this, why we use them? It initially looks like a WTF until you either think about it or someone explains it.